### ISSUE TITLE
Support "generic" JWT creation instead of explicit private key for AppAuth

### ISSUE THREAD

**Issue by davidzenisu** [PyGithub/PyGithub#3064]
Most of the APIs used by a GitHub app are accessible through the `GitHubIntegration` class ([example](https://pygithub.readthedocs.io/en/stable/examples/Authentication.html#app-authentication)) and the class only allows an `AppAuth` object for authentication ([reference](https://pygithub.readthedocs.io/en/stable/github_integration.html)) which in turn only allows an explicit private key as an input ([reference](https://pygithub.readthedocs.io/en/stable/utilities.html#github.Auth.AppAuth)).

However, the official GitHub documentation recommends the following:

> Consider storing your GitHub App's private key in a key vault [...] and making it sign-only.

Source: [Best practices for creating a GitHub App (Private Keys)](https://docs.github.com/en/apps/creating-github-apps/about-creating-github-apps/best-practices-for-creating-a-github-app#private-keys).

This is not possible with the current interfaces which all require an explicit private key.
Would it be possible considering an approach where the user passes a function as an authentication method for the `AppAuth` class as an alternative to the `private_key` parameter. This method is expected to return a JWT which could then be used for any subsequent processes (e.g. as alternative path of the `create_jwt` implementation, see [here](https://github.com/PyGithub/PyGithub/blob/c44ec523b1830c2d747fcf417b478562854664e8/github/Auth.py#L244))

**Comment by davidzenisu** [comment]
PS: Basically, the same feature request as the one made for github3.py: https://github.com/sigmavirus24/github3.py/issues/1087. Should make a concrete reference implementation more tangible.

**Comment by EnricoMi** [comment]
There is a draft to provide a sign function to AppAuth: #3064.

**Comment by EnricoMi** [comment]
@davidzenisu do you think #3065 works for you?

**Comment by davidzenisu** [comment]
Hi @EnricoMi!

Thanks a ton! I'm really sorry for getting back to you so late, this slipped through my notifications. I've already had the change to test the newly released functionality and it works exactly as expected. As far as I'm aware this library is one of the only ones out of the [officially endorsed REST API clients](https://docs.github.com/en/rest/using-the-rest-api/libraries-for-the-rest-api?apiVersion=2022-11-28) that can be used the way GitHub themselves officially recommends using it in their [best practices](https://docs.github.com/en/apps/creating-github-apps/about-creating-github-apps/best-practices-for-creating-a-github-app#private-keys).

Again, thanks a lot for implementing this!

PS: For anyone interested, I've created (a somewhat rough) [gist](https://gist.github.com/davidzenisu/a33c2344aa2bb48074f1ec61773b8ec0) that I've successfully tested on my end. Feel free to refine and expand upon it if you're in a similar scenario.

**Comment by EnricoMi** [comment]
Thanks for the praise! And thanks for sharing the gist. Would be interesting to see how this simplifies using `github.Auth.AppAuth`, as it implements most of that.

**Comment by davidzenisu** [comment]
Sorry, I actually shared the wrong gist (I've now fixed it in my previous comment):

- Gist prior to using pygithub using Azure Key Vault & httpx client: [jwt_generation_akv.py](https://gist.github.com/davidzenisu/a33c2344aa2bb48074f1ec61773b8ec0)
- Gist using Azure Key Vault & pygithub >= 2.6.0 (which I actually meant to share): [pygithub_akv.py](https://gist.github.com/davidzenisu/4d3f6b712688f8a3cc4f0b8cb212b353)

The main overhead I'm currently still running into is generating the rest of the JWT because there's no way to provide only the signed payload instead of the private key to the jwt library that I'm aware of.

**Comment by EnricoMi** [comment]
Interesting. This should further be reduced to only providing a `sign(digest) -> str` method that does the actual cryptographic signature and have the rest implemented by PyGithub `AppAuth`.

**Comment by davidzenisu** [comment]
Yes, this could be a viable approach and I would love to be able to do that.

My main concern is that the underlying library ([PyJWT](https://pyjwt.readthedocs.io/en/stable/)) doesn't support this "custom" approach of providing the signed payload to build the JWT, hence the extensive overhead/low-level implementation.

Or am I missing something here?
